flocker/control/_persistence.py

```diff
@@ -160,21 +266,65 @@ def __init__(self, reactor, path):
         """
         MultiService.__init__(self)
         self._path = path
+        self._config_path = self._path.child(b"current_configuration.json")
         self._change_callbacks = []
         LeaseService(reactor, self).setServiceParent(self)
 
     def startService(self):
         if not self._path.exists():
             self._path.makedirs()
-        self._config_path = self._path.child(b"current_configuration.v1.json")
+        self.load_configuration()
+        MultiService.startService(self)
+        _LOG_STARTUP(configuration=self.get()).write(self.logger)
+
+    def _process_v1_config(self):
+        """
+        Check if a v1 configuration file exists and upgrade it if necessary.
+        After upgrade, the v1 configuration file is retained with an archived
+        file name.
+        """
+        v1_config_path = self._path.child(
+            b"current_configuration.v1.json")
+        v1_archived_path = self._path.child(
+            b"current_configuration.v1.old.json")
+        # Check for a v1 config and upgrade to latest if found.
+        if v1_config_path.exists():
+            v1_json = v1_config_path.getContent()
+            with _LOG_UPGRADE(self.logger,
+                              configuration=v1_json,
+                              source_version=1,
+                              target_version=_CONFIG_VERSION):
+                updated_json = migrate_configuration(
+                    1, _CONFIG_VERSION, v1_json,
+                    ConfigurationMigration
+                )
+                self._config_path.setContent(updated_json)
+                v1_config_path.moveTo(v1_archived_path)
+
+    def load_configuration(self):
+        """
+        Load the persisted configuration, upgrading the configuration format
+        if an older version is detected.
+        """
+        self._process_v1_config()
         if self._config_path.exists():
-            self._deployment = wire_decode(
-                self._config_path.getContent())
+            config_json = self._config_path.getContent()
+            config_dict = loads(config_json)
+            config_version = config_dict['version']
+            if config_version < _CONFIG_VERSION:
+                with _LOG_UPGRADE(self.logger,
+                                  configuration=config_json,
+                                  source_version=config_version,
+                                  target_version=_CONFIG_VERSION):
+                    config_json = migrate_configuration(
+                        config_version, _CONFIG_VERSION,
+                        config_json, ConfigurationMigration)
+            config = wire_decode(config_json)
```

Comment:
It took me a long while to figure out how and why this code was different from `_process_v1_config`. I *guess* it's because in v1 we stored the version number in the filename rather than in the config itself, hence the need for a special case. 

Please add a comment explaining why we special-case v1. It'll probably save future readers some time.